graylog2-server/src/test/java/org/graylog/plugins/views/migrations/V20200204122000_MigrateUntypedViewsToDashboardsTest.java

```diff
@@ -0,0 +1,226 @@
+    }
+
+    @Test
+    @MongoDBFixtures("V20200204122000_MigrateUntypedViewsToDashboardsTest/typed_views.json")
+    public void doesNotChangeTypedViews() throws Exception {
+        this.migration.upgrade();
+
+        final MigrationCompleted migrationCompleted = captureMigrationCompleted();
+        assertThat(migrationCompleted.viewIds()).isEmpty();
+
+        verify(this.viewsCollection, never()).updateOne(any(), any());
+        verify(this.searchesCollection, never()).updateOne(any(), any());
+    }
+
+    @Test
+    @MongoDBFixtures("V20200204122000_MigrateUntypedViewsToDashboardsTest/mixed_typed_and_untyped_views.json")
+    public void migratesOnlyUntypedViewsIfMixedOnesArePresent() throws Exception {
+        this.migration.upgrade();
+
+        assertViewsMigrated("5c8a613a844d02001a1fd2f4");
+
+        assertSavedViews(1, resourceFile("V20200204122000_MigrateUntypedViewsToDashboardsTest/mixed_typed_and_untyped_views-views_after.json"));
+        assertSavedSearches(1, resourceFile("V20200204122000_MigrateUntypedViewsToDashboardsTest/mixed_typed_and_untyped_views-searches_after.json"));
+    }
+
+    @Test
+    @MongoDBFixtures("V20200204122000_MigrateUntypedViewsToDashboardsTest/query_with_query_string.json")
+    public void migratesQueryStringFromQueryIntoWidgetsAndSearchTypes() throws Exception {
+        this.migration.upgrade();
+
+        assertViewsMigrated("5c8a613a844d02001a1fd2f4");
+
+        assertSavedViews(1, resourceFile("V20200204122000_MigrateUntypedViewsToDashboardsTest/query_with_query_string-views_after.json"));
+        assertSavedSearches(1, resourceFile("V20200204122000_MigrateUntypedViewsToDashboardsTest/query_with_query_string-searches_after.json"));
+    }
+
+    private void assertViewsMigrated(String... viewId) {
+        final MigrationCompleted migrationCompleted = captureMigrationCompleted();
+        assertThat(migrationCompleted.viewIds()).containsExactlyInAnyOrder(viewId);
+    }
+
+    private void assertSavedViews(int count, String viewsCollection) throws Exception {
+        assertEntityCollection(count, viewsCollection, this.viewsCollection);
+    }
+
+    private void assertSavedSearches(int count, String searchesCollection) throws Exception {
+        assertEntityCollection(count, searchesCollection, this.searchesCollection);
+    }
+
+    private void assertEntityCollection(int count, String expectedCollection, MongoCollection<Document> actualCollection) throws Exception {
+        final ArgumentCaptor<Document> newEntitiesCaptor = ArgumentCaptor.forClass(Document.class);
+        verify(actualCollection, times(count)).updateOne(any(), newEntitiesCaptor.capture());
+        final List<Document> newEntities = newEntitiesCaptor.getAllValues();
+        assertThat(newEntities).hasSize(count);
+
+        JSONAssert.assertEquals(expectedCollection, toJSON(newEntities), true);
```

Comment:
I think it would make sense to extract the assertion into another private method and convert checked exceptions into `RuntimeException`s there to get rid of all the `throws` declarations.

Reply:
✅